### PR TITLE
[JUJU-2420] Use model-config instead of environment variable to control prompting globally

### PR DESCRIPTION
**Pull request by jack-w-shaw**
We decided that environment variables were a bad idea for this feature. They're a bit magic and also potentially dangerous

Since the requirement for this extra layer of protection is is property of models, add this setting as a value for the (currently unused) 'mode' config option

By default for 3.0 and 3.1 this value will be absent. However, this will change in 3.2 

The first three commits are back-ported from develop, where we abstracted out the prompting behaviour. However, this abstraction is already used by destroy-* commands which (for 3.0 and 3.1 only) behave differently. So there is still some (temporary) duplication here.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [x] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

Verify the config option `mode` is present and defaults to the empty string

for each of `remove-machine`, `remove-application` and `remove-unit` verify that:
- When `mode` doesn't contains `requires-prompts`, no prompting occurs
- When `mode` does contains `requires-prompts`, prompting occurs
  - This prompt can be removed using the `--no-prompt` flag

Verify nothing breaks for a 3.0.2 model

Verify that destroy-* commands still function as before

### CODE REVIEW

**Review by SimonRichardson — COMMENTED** [review]
Do we actually want another config-flag here? I'm not sure we do, I was thinking we could use `mode`?

**Comment by jack-w-shaw** [comment]
> Do we actually want another config-flag here? I'm not sure we do, I was thinking we could use `mode`?

How specifically would that work? Bearing in mind we need to default to no prompts for 3.0 and 3.1, then default to showing prompts in 3.1

**Comment by SimonRichardson** [comment]
Request the model-config, if the mode doesn't have "requires-prompts" then don't do anything. Make the default mode in 3.1 "require-prompts". Note: that mode is a csv, currently, the only thing that is available is "strict" mode.

The PR that introduced it https://github.com/juju/juju/pull/12183

**Comment by jack-w-shaw** [comment]
Some thought will need to be put into this for how this behaves under model migration, since there is a danger you import a model with this disabled unknowingly.

My proposal would be don't worry about this in 3.0 or 3.1, since the default is not to prompt. However, in 3.2 when you import a model display a warning that prompting will not occur, if the option is disabled

File: cmd/modelcmd/confirmation.go [hunk 1]
```diff
@@ -4,52 +4,71 @@
...
+
+// NeedsConfirmation returns indicates whether confirmation is required or not.
+func (c *RemoveConfirmationCommandBase) NeedsConfirmation(client ModelConfigAPI) bool {
+	if c.assumeNoPrompt {
+		return false
+	}
+
+	configAttrs, err := client.ModelGet()
+	if err != nil {
+		// Play it safe
```
Comment (SimonRichardson) [inline cmd/modelcmd/confirmation.go:65]: :+1:

File: environs/config/config.go [hunk 2]
```diff
@@ -1462,13 +1460,25 @@ func (c *Config) validateCharmHubURL() error {
...
 
-// Mode returns the mode type for the configuration.
-// Only two modes exist at the moment (strict or ""). Empty string
-// implies compatible mode.
-func (c *Config) Mode() ([]string, bool) {
+const (
+	// RequiresPromptsMode is used to tell clients interacting with
+	// model that confirmation prompts are required when removing
+	// potentially important resources
+	RequiresPromptsMode = "requires-prompts"
```
Comment (SimonRichardson) [inline environs/config/config.go:1467]: :+1:

**Review by SimonRichardson — APPROVED** [review]
LGTM

**Pull request merged**